ssl/statem/statem_clnt.c

```diff
@@ -1522,6 +1804,63 @@ MSG_PROCESS_RETURN tls_process_server_hello(SSL_CONNECTION *s, PACKET *pkt)
         goto err;
     }
 
+#ifndef OPENSSL_NO_ECH
+    /*
+     * If we sent an ECH then check if that worked based on the
+     * ServerHello.random confirmation trick. If that is good
+     * then we'll swap over the inner and outer contexts and
+     * proceed with inner. There are some HRR wrinkles too
+     * though.
+     */
+    if (s->ext.ech.es != NULL
+        && s->ext.ech.done != 1 && s->ext.ech.ch_depth == 0
+        && s->ext.ech.grease == OSSL_ECH_NOT_GREASE
+        && s->ext.ech.attempted_type == TLSEXT_TYPE_ech) {
+        /* check the ECH accept signal */
+        if (ech_calc_confirm(s, hrr, c_signal, shbuf, shlen) != 1
+            || ech_find_confirm(s, hrr, s_signal, shbuf, shlen) != 1
+            || memcmp(s_signal, c_signal, 8) != 0) {
+            OSSL_TRACE(TLS, "ECH accept check failed\n");
+# ifdef OSSL_ECH_SUPERVERBOSE
+            ech_pbuf("ECH client accept val:", c_signal, 8);
+            ech_pbuf("ECH server accept val:", s_signal, 8);
+# endif
+            s->ext.ech.success = 0;
+        } else { /* match, ECH worked */
+            OSSL_TRACE_BEGIN(TLS) {
+                BIO_printf(trc_out, "ECH accept check ok\n");
+                BIO_printf(trc_out, "ECH set session hostname to %s\n",
+                           s->ext.hostname ? s->ext.hostname : "NULL");
+            } OSSL_TRACE_END(TLS);
+            s->ext.ech.success = 1;
+        }
+        if (!hrr && s->ext.ech.success == 1) {
+            if (ech_swaperoo(s) != 1
+                || ech_make_transcript_buffer(s, hrr, shbuf, shlen, &abuf, &alen,
+                                              &chend, &fixedshbuf_len) != 1
+                || ech_reset_hs_buffer(s, abuf, alen) != 1) {
+                OPENSSL_free(abuf);
+                SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_INTERNAL_ERROR);
+                goto err;
+            }
+            OPENSSL_free(abuf);
+        } else if (!hrr) {
+            /*
+             * If we got retry_configs then we should be validating
+             * the outer CH, so we better set the hostname for the
+             * connection accordingly.
+             */
+            s->ext.ech.former_inner = s->ext.hostname;
+            s->ext.hostname = NULL;
+            if (s->ext.ech.outer_hostname != NULL) {
+                s->ext.hostname = OPENSSL_strdup(s->ext.ech.outer_hostname);
+                if (s->ext.hostname == NULL)
+                    return 0;
```

Comment:
Missing SSLfatal() and also should `goto err;`, shouldn't it?

Reply:
added and yes (though `goto err` doesn't make a difference here as `extensions` weren't allocated yet, but it's still better yes.